### PR TITLE
Immediately evict pods and delete node when cloud says node is gone.

### PR DESCRIPTION
**Pull request by cjcullen**
Let's try this again... (previous PR #21187).

@gmarek @alex-mohr @lavalamp

### CODE REVIEW

File: pkg/controller/node/nodecontroller.go [hunk 1]
```diff
@@ -496,6 +488,43 @@ func (nc *NodeController) monitorNodeStatus() error {
 	return nil
 }
 
+var nodeExistsInCloudProvider = func(nc *NodeController, nodeName string) (bool, error) {
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Why not ordinary NodeController method?

File: pkg/controller/node/nodecontroller.go [hunk 2]
```diff
@@ -496,6 +488,43 @@ func (nc *NodeController) monitorNodeStatus() error {
...
+// forcefullyDeleteNode immediately deletes all pods on the node, and then
+// deletes the node itself.
+func (nc *NodeController) forcefullyDeleteNode(nodeName string) error {
+	selector := fields.OneTermEqualSelector(client.PodHost, nodeName)
+	options := api.ListOptions{FieldSelector: selector}
+	pods, err := nc.kubeClient.Core().Pods(api.NamespaceAll).List(options)
+	if err != nil {
+		return fmt.Errorf("unable to list pods on node %q: %v", nodeName, err)
+	}
+	for _, pod := range pods.Items {
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: I'm not quite sure how it'll interact with mirror pods... @yujuhong

Comment (yujuhong) [inline pkg/controller/node/nodecontroller.go]: If kubelet is semi-functioning, it may try to recreate the mirror pods. You'll be left with running pods on a node that doesn't exist anymore. Since this PR deletes pods/nodes only when the vm has disappeared, it should be fine.

File: pkg/controller/node/nodecontroller.go (hunk 1 above)
Comment (cjcullen) [inline pkg/controller/node/nodecontroller.go]: I did it this way for testability. I could also made it a function pointer in the struct. I lean towards stubbable package functions for testing, but I'll change it if you'd like.

Comment (lavalamp) [inline pkg/controller/node/nodecontroller.go]: function pointer in struct is definitely better. Also define the actual function with a name. Once a test clobbers this, it'll be clobbered for any other test.

File: pkg/controller/node/nodecontroller.go [hunk 3]
```diff
@@ -462,33 +462,25 @@ func (nc *NodeController) monitorNodeStatus() error {
...
 			// Check with the cloud provider to see if the node still exists. If it
-			// doesn't, delete the node and all pods scheduled on the node.
+			// doesn't, delete the node immediately.
 			if readyCondition.Status != api.ConditionTrue && nc.cloud != nil {
-				instances, ok := nc.cloud.Instances()
-				if !ok {
-					glog.Errorf("%v", ErrCloudInstance)
+				exists, err := nodeExistsInCloudProvider(nc, node.Name)
+				if err != nil {
+					glog.Errorf("%v", err)
```
Comment (lavalamp) [inline pkg/controller/node/nodecontroller.go]: "Error determining if node %v exists in cloud: %v"

File: pkg/controller/node/nodecontroller.go [hunk 4]
```diff
@@ -462,33 +462,25 @@ func (nc *NodeController) monitorNodeStatus() error {
...
-						glog.V(2).Infof("Deleting node %s is delayed while pods are evicted", node.Name)
-						nc.evictPods(node.Name)
-						continue
-					}
-
-					if err := nc.kubeClient.Core().Nodes().Delete(node.Name, nil); err != nil {
-						glog.Errorf("Unable to delete node %s: %v", node.Name, err)
-						continue
-					}
+					go func(nodeName string) {
```
Comment (lavalamp) [inline pkg/controller/node/nodecontroller.go]: defer util.HandleCrash()

File: pkg/controller/node/nodecontroller_test.go [hunk 5]
```diff
@@ -451,6 +455,67 @@ func TestMonitorNodeStatusEvictPods(t *testing.T) {
...
+	for i := 0; i < tries; i++ {
+		if nodeDeleted = func() bool {
+			fnh.deleteLock.Lock()
+			defer fnh.deleteLock.Unlock()
+			return len(fnh.DeletedNodes) == 1 && fnh.DeletedNodes[0].Name == "node0"
+		}(); nodeDeleted {
+			break
+		}
+		// Deletion happens in a goroutine, so it may not have completed yet.
+		time.Sleep(1 * time.Millisecond)
```
Comment (lavalamp) [inline pkg/controller/node/nodecontroller_test.go]: This is racy. Figure out a way to test that's not racy. :)

File: pkg/controller/node/nodecontroller_test.go [hunk 6]
```diff
@@ -451,6 +455,67 @@ func TestMonitorNodeStatusEvictPods(t *testing.T) {
 	}
 }
 
+// TestCloudProviderNoRateLimit tests that monitorNodes() immediately deletes
+// pods and the node when kubelet has not reported, and the cloudprovider says
+// the node is gone.
+func TestCloudProviderNoRateLimit(t *testing.T) {
+	tmp := nodeExistsInCloudProvider
```
Comment (lavalamp) [inline pkg/controller/node/nodecontroller_test.go]: Please don't test things like this. If you absolutely must, you *have* to restore the value in a defer block (because t.Failf exits the test immediately).

File: pkg/controller/node/nodecontroller_test.go [hunk 7]
```diff
@@ -451,6 +455,67 @@ func TestMonitorNodeStatusEvictPods(t *testing.T) {
...
+	nodeExistsInCloudProvider = func(nc *NodeController, nodeName string) (bool, error) {
```
Comment (lavalamp) [inline pkg/controller/node/nodecontroller_test.go]: But in this case, there's no good reason for this to not be a member of node controller.

Comment (cjcullen) [inline pkg/controller/node/nodecontroller_test.go]: Done.

File: pkg/controller/node/nodecontroller_test.go (hunk 6 above)
Comment (cjcullen) [inline pkg/controller/node/nodecontroller_test.go]: Done.

File: pkg/controller/node/nodecontroller_test.go (hunk 5 above)
Comment (cjcullen) [inline pkg/controller/node/nodecontroller_test.go]: Done.

File: pkg/controller/node/nodecontroller.go (hunk 1 above)
Comment (cjcullen) [inline pkg/controller/node/nodecontroller.go]: Changed to a function pointer.

File: pkg/controller/node/nodecontroller.go (hunk 4 above)
Comment (cjcullen) [inline pkg/controller/node/nodecontroller.go]: Done.

File: pkg/controller/node/nodecontroller.go (hunk 3 above)
Comment (cjcullen) [inline pkg/controller/node/nodecontroller.go]: Done.

File: pkg/controller/node/nodecontroller_test.go [hunk 8]
```diff
@@ -451,6 +458,58 @@ func TestMonitorNodeStatusEvictPods(t *testing.T) {
...
+	nodeController.nodeExistsInCloudProvider = func(nodeName string) (bool, error) {
+		return false, nil
+	}
+	// monitorNodeStatus should allow this node to be immediately deleted
+	if err := nodeController.monitorNodeStatus(); err != nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+	select {
+	case <-fnh.deleteWaitChan:
+	case <-time.After(5 * time.Second):
```
Comment (lavalamp) [inline pkg/controller/node/nodecontroller_test.go]: any reason for this timeout? Why not just wait forever?

Comment (bprashanth) [inline pkg/controller/node/nodecontroller_test.go]: Fwiw I think a lot of tests use wait.ForeverTestTimeout in similar situations, because the unittest will auto timeout with a weird stacktrace in 5m anyway and this error message is more helpful.

Comment (cjcullen) [inline pkg/controller/node/nodecontroller_test.go]: Ahh, I didn't know about that. Changed to wait.ForeverTestTimeout.

File: pkg/controller/node/nodecontroller.go [hunk 9]
```diff
@@ -496,6 +491,43 @@ func (nc *NodeController) monitorNodeStatus() error {
...
+	options := api.ListOptions{FieldSelector: selector}
+	pods, err := nc.kubeClient.Core().Pods(api.NamespaceAll).List(options)
+	if err != nil {
+		return fmt.Errorf("unable to list pods on node %q: %v", nodeName, err)
+	}
+	for _, pod := range pods.Items {
+		if pod.Spec.NodeName != nodeName {
+			continue
+		}
+		if err := nc.forcefullyDeletePod(&pod); err != nil {
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Let me rephrase my question (I'd love to attach it to 'forcefullyDeletePod' function, but github...): how does force-deleting (GracePeriodSeconds=0) work on mirror pods and/or daemon pods. @yujuhong @mikedanese. Will it correctly delete it from the API server and it won't come back, or they'll get recreated and we'll have a race condition here?

File: pkg/controller/node/nodecontroller_test.go [hunk 10]
```diff
@@ -451,6 +459,58 @@ func TestMonitorNodeStatusEvictPods(t *testing.T) {
...
+		return false, nil
+	}
+	// monitorNodeStatus should allow this node to be immediately deleted
+	if err := nodeController.monitorNodeStatus(); err != nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+	select {
+	case <-fnh.deleteWaitChan:
+	case <-time.After(wait.ForeverTestTimeout):
+		t.Errorf("Timed out waiting 5 seconds for node to be deleted")
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller_test.go]: s/5 seconds/%v/, s/")/", wait.ForeverTestTimeout)/

File: pkg/controller/node/nodecontroller.go (hunk 9 above)
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Sorry - this has nothing to do with this PR. It's a problem for 'normal' Node eviction, but in your case it's OK.

**Comment by gmarek** [comment]
One nit and LGTM.

File: pkg/controller/node/nodecontroller_test.go (hunk 10 above)
Comment (cjcullen) [inline pkg/controller/node/nodecontroller_test.go]: Done.

**Comment by gmarek** [comment]
@k8s-bot test this issue: #22603
